### ISSUE TITLE
Converting graph information from dgl.Graph back into Structure

### ISSUE THREAD

**Issue by mwolinska** [materialsvirtuallab/matgl#139]
Hi, 

I am looking to extract the gradient with respect to atom positions from the band gap prediction model "MEGNet-MP-2019.4.1-BandGap-mfi".

By going through the wrapper into MEGNet.predict_structure function I can extract the gradients with respect to the edge attributes. Is there a way to convert this information into gradient with respect to atom positions in the structure (i.e. g.ndata["pos"] in the code below).

Below I provide the unpacked method to access the gradient within the `predict_structure()` function. 
I can do the same within the `forward` method, but I am unclear which representation of the graph will provide me with the information I require for translation back to the structure.

Thank you for your help. 

Code to reproduce: 
```python
import matgl
import torch

from mp_api.client import MPRester
from pymatgen.io.ase import AseAtomsAdaptor
from matgl.ext.pymatgen import Structure2Graph
from matgl.graph.compute import compute_pair_vector_and_distance

if __name__ == '__main__':
    # loading arbitrary structure from Materials Project
    with MPRester() as mpr:
        structure = mpr.get_structure_by_material_id("mp-1840", final=True)
    atoms_2 = AseAtomsAdaptor.get_atoms(structure)
   
    # selecting a band gap method
    state_feats = torch.tensor([3])

    band_gap_model_wrapper = matgl.load_model("MEGNet-MP-2019.4.1-BandGap-mfi")

    # unpacking forward method of MEGNet.predict_structure to access gradient information
    graph_converter = Structure2Graph(
        element_types=band_gap_model_wrapper.model.element_types,
        cutoff=band_gap_model_wrapper.model.cutoff,
    )
    g, state_feats_default = graph_converter.get_graph(structure)
    if state_feats is None:
        state_feats = torch.tensor(state_feats_default)
    bond_vec, bond_dist = compute_pair_vector_and_distance(g)
    g.edata["edge_attr"] = band_gap_model_wrapper.model.bond_expansion(bond_dist)

    # adding requires_grad argument edges to access gradients
    g.edata["edge_attr"].requires_grad_()

    model_output = band_gap_model_wrapper.model(g, g.edata["edge_attr"], g.ndata["node_type"], state_feats)

    gradient_wrt_edge_attr = torch.autograd.grad(
        model_output, g.edata["edge_attr"], create_graph=True, retain_graph=True,
    )
    
    model_output = model_output.detach()
    model_output_converted = band_gap_model_wrapper.transformer.inverse_transform(model_output)

    # asserting that unpacked method gives the same answer as the wrapper method provided
    band_gap_value = band_gap_model_wrapper.predict_structure(
        structure=structure,
        state_feats=state_feats
    )

    assert model_output_converted == band_gap_value

```

**Comment by mwolinska** [comment]
Hi, thank you for your quick reply!

I'm not sure if I understood you correctly, but my problem isn't coming from matching a prediction to a graph. I would like to use the gradient from the model (which is with respect to the graph) to inform a change in the structure itself. Is that possible?

**Comment by mwolinska** [comment]
Perhaps the title of my issue is misleading, let me elaborate on the problem briefly.
Currently the process to make the prediction is:
1. Pass structure in
2. Convert structure to graph
3. Make a prediction on the graph 

What I would like to do is after step 3:
4. Extract gradient from model,
5. Convert gradient from the embedding representation into the gradient on each atom position. 

I think this is similar to what is done to extract the forces when computing energy, except I would like to do it when making a prediction using "MEGNet-MP-2019.4.1-BandGap-mfi".

**Comment by shyuep** [comment]
That's the same thing. In the PES models, both energy and forces (the gradient of the energies) are explicitly handled as part of the prediction. So you do have the actual structure gradient in this case.

But for property prediction models, you are predicting the band gap. You have the gradient of the band gap wrt to the graph. While you can "move" the graph in the direction of the gradient, that is no easy way to translate the modified graph back to a structure. 

Note that the gradient of band gap wrt to graph is not a force per se. It is not like the PES model where the gradient is an actual force on each atom that you can use to move the atoms using a Verlet algorithm.

**Comment by kenko911** [comment]
Hi @mwolinska. Thank you for your questions. @shyuep is right that we only return the gradient of bandgap with respect to graph attributes include nodes, edges and states for property models. If you really want to get the gradient of bandgap with respect to atomic positions. I would suggest you explicitly calculating the gradient of bond expansion (edge attributes) with respect to atomic position and then perform matrix multiplication with the gradient of band gap with respect to bond expansion (edge attributes).

**Comment by mwolinska** [comment]
Hi @kenko911, thank you for your suggestion! Do you think computing the gradient of bond expansion with respect to atomic position from your implementation would be fairly straightforward?

**Comment by kenko911** [comment]
Based on the script you provided, I am confident that you can figure it out by your own ;). I will close this issue if no further question. Thanks.
